---
title: "Method_description"
output: rmarkdown::github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
load('./data/ST_heart_08032018.RData')
unique.regions<-data.frame(col = unique(heart$atlas$color2), acronym = na.omit(unique(heart$atlas$acronym)), name = na.omit(unique(heart$atlas$name)) )
nuclei.to.transcripts<-cor.test(heart$atlas$nuclei, apply(heart$genes,1,sum))

p.value.report<-function(p.value){
  if(p.value < 0.0001)
    return('0.0001')
  if(p.value < 0.001)
    return('0.001')
  if(p.value < 0.01)
    return('0.01')
  if(p.value < 0.05)
    return('0.05')
  if(p.value > 0.05)
    return(round(p.value, 2))
}

library(mclust)
nuclei.dataframe<-data.frame(cell.nuclei = heart$atlas$nuclei, transcripts =  apply(heart$genes,1,sum)) 
clusters<-Mclust(nuclei.dataframe, G=2)

```

## Results

### Reconstructing the transcriptome in 3D.

From a total of nine sections (_n_ = `r length(unique(heart$atlas$image))`) we reconstructed the embrionic heart in three-dimensions by registering individual tissue sections onto matched planes within a 3D reference atlas that we constructed based on (). We further deliminated the embrionic heart into `r length(unique(heart$atlas$acronym))` unique anatomical regions (`r paste(unique.regions[,2], unique.regions[,3], sep=': ')`). 

Using cell segmentation algorithms on the HNE image we detected a total of nuclei in the tissue. Out of all the cells detected a total of XXX were captured by one of the barcoded RNA-capturing spots (`r as.integer(sum(heart$atlas$nuclei))` out of cell nuclei) with an average of `r as.integer(mean(heart$atlas$nuclei))` nuclei per spot (_M_ = `r round(mean(heart$atlas$nuclei), 2)` nuclei, _SD_ = `r round(sd(heart$atlas$nuclei), 2)` nuclei). A cross the entire tissue there was a negative correlation between number of cell nuclei in each spot and the number of transcripts detected in the same spot _r_ = `r round(nuclei.to.transcripts$estimate, 2)` (_t_ = `r round(nuclei.to.transcripts$statistic, 2)`, _df_ = `r round(nuclei.to.transcripts$parameter, 2)`, _p_ < `r p.value.report(nuclei.to.transcripts$p.value)`). However, upon closer examination, cell density and number of transcripts seggregated into two seperate bivariate distributions: one with an average of cell nuclei per spots and another distribution with an average of less than ten cell nuclei per spot. These two clusters were located into different anatomical compartments where spots with an average number of cell nuclei .

```{r  fig1, include=FALSE, message=FALSE, warning=FALSE, echo=FALSE}
quartz(width=6.908108, height= 2.5)
par(mar=c(4,6,1,0), yaxs='i', mfrow=c(1,2))
plot(heart$atlas$nuclei+1, apply(heart$genes,1,sum), pch=16, cex=0.5, col=clusters$classification, xlab='', ylab='', ylim=c(0,30000), xlim=c(0,100),axes = FALSE)
axis(1, at=c(0,50,100))
axis(2, las=1, at=c(0,15000,30000))
mtext('Transcript count in spot', 2, 4)
mtext('Cell nuclei in spot', 1, 3)
text(70, 25000, 'cluster 2')
text(70, 29000, 'cluster 1', col='red')
par(mar=c(4,4,1,1))
plot(rnorm(100))
quartz.save('./images/nuclei_to_count.png', type='png', dpi=300)
```
![Showing spots in 3D.](./images/nuclei_to_count.png)


## Methods

### Reference atlas construction.

We 

### Segmentation.

### Registration.

